oauth/scopes.js

```diff
@@ -0,0 +1,613 @@
+   * `A.filtered(B)`. It returns a new `ScopeSet` object.
+   *
+   */
+  difference(other) {
+    other = coerce(other);
+    const result = new ScopeSet();
+    this._iterScopes((scope, implicants) => {
+      if (! other._hasSomeScope(implicants)) {
+        result._addScope(scope, implicants);
+      }
+    });
+    return result;
+  }
+
+  /**
+   * Get the combined set of scope values implied by either this `ScopeSet`
+   * or another.
+   *
+   * `A.union(B)` returns the smallest set of scope values that are implied
+   * either `A` or `B` (or both).  It returns a new `ScopeSet` object.
+   *
+   */
+  union(other) {
+    other = coerce(other);
+    const result = new ScopeSet();
+    // Take scopes from `this` that are not implied by a scope from `other.
+    this._iterScopes((scope, implicants) => {
+      if (! other._hasSomeScope(implicants)) {
+        result._addScope(scope, implicants);
+      }
+    });
+    // Combine those with any scopes from `other` that are
+    // not implied by a strictly-more-general scope from `this`.
+    other._iterScopes((scope, implicants) => {
+      if (! this._hasSomeScope(implicants) || this._hasScope(scope)) {
+        result._addScope(scope, implicants);
+      }
+    });
+    return result;
+  }
+
+}
+
+
+/**
+ * A little helper function for coercing different
+ * kinds of input data into a `ScopeSet` instance.
+ *
+ */
+function coerce(scopes) {
+  if (scopes instanceof ScopeSet) {
+    return scopes;
+  }
+  if (typeof scopes === 'string') {
+    return new ScopeSet([scopes]);
+  }
+  return new ScopeSet(scopes);
+}
+
+
+/**
+ * An iterator yielding all implicants of the given scope value.
+ *
+ */
+function getImplicantValues(value) {
+  if (value.startsWith('https:')) {
+    return getImplicantValuesForURLScope(value);
+  } else {
+    return getImplicantValuesForShortScope(value);
+  }
+}
+
+
+/**
+ * Our "short-name scopes" are things like:
+ *
+ *   * "openid"
+ *   * "profile"
+ *   * "profile:write"
+ *   * "profile:display_name:write"
+ *
+ * The scope value is a colon-separated list of
+ * alphanumeric strings.  The special value "write"
+ * may appear as the last component to indicate write
+ * access, otherwise access is read-only.
+ *
+ * Implication is essentially the prefix relationship
+ * on the name components, with the additional rule
+ * only "write" scopes can imply other "write" scopes.
+ *
+ */
+function* getImplicantValuesForShortScope(value) {
+  if (! VALID_SCOPE_VALUE.test(value)) {
+    throw new Error('Invalid scope value: ' + value);
+  }
+  // Parse it into its colon-separated name components,
+  // and handle the special "write" flag if present.
+  let hasWrite = false;
+  const names = value.split(':');
+  names.forEach(name => {
+    if (! VALID_SHORT_NAME_VALUE.test(name)) {
+      throw new Error('Invalid scope value: ' + value);
+    }
+  });
+  if (names[names.length - 1] === 'write') {
+    hasWrite = true;
+    names.pop();
+    if (names.length === 0) {
```

Comment:
This ensures the value of `write` without a scope name is disallowed?

Reply:
Yes, I'll add a comment to that effect.